File: services/analytics-api/src/trackbus_analytics/forecast.py
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Literal


@dataclass(frozen=True)
class ForecastInput:
    recent_occupancy: Sequence[int]
    capacity: int
    hour: int
    day_type: Literal["weekday", "weekend"] = "weekday"
    weather: Literal["clear", "rain"] = "clear"
    event_nearby: bool = False


@dataclass(frozen=True)
class ForecastResult:
    expected_occupancy: int
    expected_percent: int
    lower_bound: int
    upper_bound: int
    confidence: float


def forecast_occupancy(data: ForecastInput) -> ForecastResult:
    """Transparent baseline; future ML models must outperform it in evaluation."""
    if data.capacity <= 0:
        raise ValueError("capacity must be positive")
    history = [max(0, int(value)) for value in data.recent_occupancy[-6:]]
    if not history:
        raise ValueError("recent_occupancy must contain at least one observation")

    weights = list(range(1, len(history) + 1))
    baseline = sum(value * weight for value, weight in zip(history, weights)) / sum(weights)
    rush_hour = data.day_type == "weekday" and (7 <= data.hour <= 9 or 17 <= data.hour <= 20)
    estimate = baseline * (1.13 if rush_hour else 1)
    estimate *= 1.07 if data.weather == "rain" else 1
    estimate *= 1.10 if data.event_nearby else 1
    expected = min(data.capacity, max(0, round(estimate)))
    spread = max(3, round(data.capacity * 0.08))

    return ForecastResult(
        expected_occupancy=expected,
        expected_percent=round(expected / data.capacity * 100),
        lower_bound=max(0, expected - spread),
        upper_bound=min(data.capacity, expected + spread),
        confidence=0.86 if len(history) >= 6 else 0.68,
    )

File: lib/forecast.ts
export type ForecastInput = {
  recentOccupancy: number[];
  capacity: number;
  hour: number;
  dayType: "weekday" | "weekend";
  weather?: "clear" | "rain";
  eventNearby?: boolean;
};

export type ForecastResult = {
  expectedOccupancy: number;
  expectedPercent: number;
  lowerBound: number;
  upperBound: number;
  confidence: number;
};

/**
 * Transparent baseline for the first pilot. It is intentionally simple and
 * measurable; later ML models must beat it before being promoted.
 */
export function forecastOccupancy(input: ForecastInput): ForecastResult {
  const history = input.recentOccupancy.slice(-6);
  const weighted = history.reduce(
    (sum, value, index) => sum + value * (index + 1),
    0,
  );
  const divisor = history.reduce((sum, _, index) => sum + index + 1, 0) || 1;
  const base = weighted / divisor;
  const rushHourFactor =
    input.dayType === "weekday" &&
    ((input.hour >= 7 && input.hour <= 9) || (input.hour >= 17 && input.hour <= 20))
      ? 1.13
      : 1;
  const weatherFactor = input.weather === "rain" ? 1.07 : 1;
  const eventFactor = input.eventNearby ? 1.1 : 1;
  const expected = Math.min(
    input.capacity,
    Math.max(0, Math.round(base * rushHourFactor * weatherFactor * eventFactor)),
  );
  const spread = Math.max(3, Math.round(input.capacity * 0.08));

  return {
    expectedOccupancy: expected,
    expectedPercent: Math.round((expected / input.capacity) * 100),
    lowerBound: Math.max(0, expected - spread),
    upperBound: Math.min(input.capacity, expected + spread),
    confidence: history.length >= 6 ? 0.86 : 0.68,
  };
}
